akka-stream/src/main/scala/akka/stream/impl/fusing/FoldWithin.scala

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.stream.stage._
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+
+import scala.concurrent.duration._
+
+/**
+ * This is a more scalable and general case of [[GroupedWeightedWithin]]
+ * which groups a stream into vectors based on custom weight and time.
+ * The problem with that solution is each grouped vector must fit into memory before emitting to the next stage.
+ * That won't work for the use case of writing large files.
+ * The desirable behavior is to write data as they come as opposed to accumulate everything until the group slicing condition is met.
+ * In this case, the output channel needs to be closed if there is no data arriving within certain time to avoid connection timeout.
+ * This custom flow uses custom aggregator to support such use cases.
+ * Upstream inputs are continuously aggregated as they arrive.
+ * The aggregator/grouping can be terminated and emitted based on custom coditions
+ *
+ * @param seed        initiate the aggregated output with first input
+ * @param aggregate   sequentially aggregate input
+ * @param emitOnAgg   decide whether the current aggregator can be emitted, invoked after each aggregate
+ * @param emitOnTimer decide whether the current aggregator can be emitted, invoked after each timer event
+ * @param harvest     this is invoked as soon as all conditions are met before emitting to next stage
+ *                    there can be extra time between harvest and next stage receiving the emitted output
+ *                    time sensitive operations can be added here, such as closing an output channel
+ */
+class FoldWith[In, Agg, Out](
```

Comment:
Maybe BoundedFold, or FoldWithBoundary, etc.
Aggregate normally is in the context of group-by.
I'm actually thinking about changing the parameter name `aggregate` to just `f`, so it can better relate to the fold semantics.

Reply:
renamed to `AggregateWithBoundary`